File: lec1/examples/linux-c/copy.c
// using read/write to copy input to output.
#include<stdio.h>
#include<unistd.h>
#include<fcntl.h>

#define BUFSIZE 512

int main(int argc, char* argv[])
{
    int fd1, fd2, n;
    char buf[BUFSIZE];

    if(argc != 3)
    {
        fprintf(stderr, "usage: copy file1 file2\n");
        return 1;
    }

    // open file1 for reading
    if((fd1 = open(argv[1], O_RDONLY)) < 0)
    {
        fprintf(stderr, "cannot open %s\n", argv[1]);
        return 1;
    }

    // open/create file2 for writing
    if((fd2 = open(argv[2], O_WRONLY | O_CREAT | O_TRUNC, 0644)) < 0)
    {
        fprintf(stderr, "cannot open %s\n", argv[2]);
        return 1;
    }

    // copy file1 to file2
    while((n = read(fd1, buf, BUFSIZE)) > 0)
        write(fd2, buf, n);

    close(fd1);
    close(fd2);
    return 0;
}


// The code is fairly simple. The first thing is to open the files. The first file is opened for reading, and the second file is opened for writing. If the second file does not exist, it is created. If the file does exist, it is truncated to zero length. The while loop reads from the first file and writes to the second file. The read() function returns the number of bytes read. If it returns zero, the end of file has been reached. If it returns a negative number, an error has occurred. The second file is closed, then the program exits.

File: lec1/examples/linux-rust/src/copy.rs
// using read/write to copy input to output.
use std::env;
use std::fs::File;
use std::io::{BufReader, BufWriter, Result, Write, Read};

const BUFSIZE: usize = 512;

fn main() -> Result<()> {
    let args: Vec<String> = env::args().collect();
    if args.len() != 3 {
        eprintln!("usage: copy file1 file2");
        std::process::exit(1);
    }

    // open file1 for reading
    let file1 = File::open(&args[1])?;
    let mut reader = BufReader::new(file1);

    // open/create file2 for writing
    let file2 = File::create(&args[2])?;
    let mut writer = BufWriter::new(file2);

    // copy file1 to file2
    let mut buf = [0u8; BUFSIZE];
    loop {
        let n = reader.read(&mut buf)?;
        if n == 0 {
            break;
        }
        writer.write_all(&buf[..n])?;
    }

    Ok(())
}

// 以下是相应的Rust代码实现，使用了标准库中的File和BufReader/BufWriter来进行文件读写操作：